zsh-history-substring-search.zsh

```diff
@@ -233,20 +233,33 @@ _history-substring-search-begin() {
     _history_substring_search_query=$BUFFER
 
     #
-    # $BUFFER contains the text that is in the command-line currently.
-    # we put an extra "\\" before meta characters such as "\(" and "\)",
-    # so that they become "\\\(" and "\\\)".
+    # compose search pattern
     #
-    local escaped_query=${BUFFER//(#m)[\][()|\\*?#<>~^]/\\$MATCH}
+    # (q)        - escape special characters within string
+    # (s:CHAR>:) - split string by CHAR
+    # j:CHAR:)   - join array to string with CHAR as seperator
+    # =          - split string in arguments
 
+    if [ "$HISTORY_SUBSTRING_SEARCH_METHOD" = 'fuzzy_words' ]; then
+      # find history by words in given order
+      _history_substring_search_query_parts=(${=_history_substring_search_query})
+    elif [ "$HISTORY_SUBSTRING_SEARCH_METHOD" = 'fuzzy_chars' ]; then
+      # find history by chars in given order
+      _history_substring_search_query_parts=(${(s::)_history_substring_search_query})
+    else
+      # find history by query
+      _history_substring_search_query_parts=(${_history_substring_search_query})
+    fi
+    local search_pattern=${(j:*:)${(q)_history_substring_search_query_parts}}
```

Comment:
I think it is but you might not want to use it (you don't want to escape the joining `*` characters), depends on the ordering, and needs to be checked.